File: linkedlist/linkedlist.js
/*
 Note. All functions are in-order

 ** Singly Linked List 
 ** New - new LinkedList(value)
 ** Insert (at start) - .prepend(value)
 ** Insert (at end) - .append(value)
 ** Insert (at middle) - .insert(index, value)
 ** Lookup - .get(index)
 ** Remove - .remove(index)
 */

// =============================================

class LinkedList {
  constructor(value) {
    this.initialize(value);
  }

  initialize = (value) => {
    this.head = { value, next: null };
    this.tail = this.head;
    this.length = 1;
  };

  isEmpty = () => {
    if (!this.head) {
      console.log('List is empty');
      return true;
    }
    return false;
  };

  append = (value) => {
    if (this.isEmpty()) {
      this.initialize(value);
      return this;
    }

    const node = { value, next: null };
    this.tail.next = node;
    this.tail = node;
    this.length++;
    return this;
  };

  prepend = (value) => {
    if (this.isEmpty()) {
      this.initialize(value);
      return this;
    }

    const node = { value, next: null };
    node.next = this.head;
    this.head = node;
    this.length++;
    return this;
  };

  print = () => {
    if (!this.isEmpty()) {
      let curr = this.head;
      let str = curr.value + '';
      while (curr.next != null) {
        curr = curr.next;
        str += ` --> ${curr.value}`;
      }
      console.log(str);
    }
  };

  traverse = (index) => {
    if (this.isEmpty()) {
      return { follower: null, curr: null, leader: null };
    }

    let follower = null;
    let curr = this.head;

    for (let i = 0; i < index && i < this.length - 1; i++) {
      follower = curr;
      curr = curr.next;
    }

    return { follower, curr, leader: curr ? curr.next : null };
  };

  get = (index) => {
    if (!this.isEmpty()) {
      return this.traverse(index).curr.value;
    }
    return null;
  };

  insert = (index, value) => {
    if (this.isEmpty()) {
      this.initialize(value);
      return this;
    }
    const node = { value, next: null };

    const { _, curr, leader } = this.traverse(index);
    curr.next = node;
    node.next = leader;
    this.length++;
    return this;
  };

  remove = (index) => {
    if (!this.isEmpty()) {
      const { follower, curr, leader } = this.traverse(index);
      if (!follower) {
        this.head = leader;
        curr.next = null;
      } else if (!leader) {
        this.tail = follower;
        follower.next = null;
      } else {
        follower.next = leader;
        curr.next = null;
      }
    }

    return this;
  };

  reverse = () => {
    if (!this.isEmpty() && this.length > 1) {
      if (this.length === 2) {
        this.tail = this.head;
        this.head = this.tail.next;
        this.head.next = this.tail;
        this.tail.next = null;
        return;
      }

      const holder = this.head;
      this.head = this.tail;
      this.tail = holder;

      let follower = this.tail;
      let curr = follower.next;

      while (curr !== null) {
        const leader = curr.next;
        curr.next = follower;
        follower = curr;
        curr = leader;
      }

      this.tail.next = null;
    }
  };
}

const ll = new LinkedList(1);
ll.append(2);
ll.append(3);
ll.append(4);
ll.append(5);
ll.print();
ll.reverse();
ll.print();

File: linkedlist/linkedlist.go
/*
 Note. All functions are in-order

 ** Singly Linked List
 ** New - new linkedlist{} and .initialize(value)
 ** Insert (at start) - .prepend(value)
 ** Insert (at end) - .append(value)
 ** Insert (at middle) - .insert(index, value)
 ** Lookup - .get(index)
 ** Remove - .remove(index)
*/

// =============================================

package main

import "fmt"

type node struct {
	value int
	next  *node
}

type linkedlist struct {
	head   *node
	tail   *node
	length int
}

func (ll *linkedlist) initialize(value int) {
	n := node{value: value, next: nil}
	ll.head = &n
	ll.tail = &n
	ll.length = 1
}

func (ll linkedlist) isEmpty() bool {
	if ll.head == nil {
		fmt.Println("List is empty")
		return true
	}
	return false
}

func (ll *linkedlist) append(value int) {
	if ll.isEmpty() {
		ll.initialize(value)
		return
	}

	n := node{value: value, next: nil}
	ll.tail.next = &n
	ll.tail = &n
	ll.length++
}

func (ll *linkedlist) prepend(value int) {
	if ll.isEmpty() {
		ll.initialize(value)
		return
	}

	n := node{value: value, next: nil}
	n.next = ll.head
	ll.head = &n
	ll.length++
}

func (ll *linkedlist) print() {
	if !ll.isEmpty() {
		curr := ll.head
		nodes := []int{curr.value}
		curr = curr.next
		for curr != nil {
			nodes = append(nodes, (*curr).value)
			curr = curr.next
		}
		fmt.Println(nodes)
	}
}

func (ll *linkedlist) traverse(index int) (*node, *node, *node) {
	if ll.isEmpty() {
		return nil, nil, nil
	}

	var follower *node
	curr := ll.head

	for i := 0; i < index && i < ll.length-1; i++ {
		follower = curr
		curr = curr.next
	}

	return follower, curr, curr.next
}

func (ll *linkedlist) get(index int) (int, bool) {
	if !ll.isEmpty() {
		_, curr, _ := ll.traverse(index)
		return curr.value, true
	}

	return 0, false
}

func (ll *linkedlist) insert(index, value int) {
	if ll.isEmpty() {
		ll.initialize(value)
		return
	}
	n := node{value: value, next: nil}

	_, curr, leader := ll.traverse(index)
	curr.next = &n
	n.next = leader
	ll.length++
}

func (ll *linkedlist) remove(index int) {
	if !ll.isEmpty() {
		follower, curr, leader := ll.traverse(index)
		if follower == nil {
			ll.head = leader
			curr.next = nil
		} else if leader == nil {
			ll.tail = follower
			follower.next = nil
		} else {
			follower.next = leader
			curr.next = nil
		}
		ll.length--
	}
}

func main() {
	ll := linkedlist{}
	ll.print()
	ll.initialize(5)
	ll.print()
	ll.append(16)
	ll.print()
	ll.prepend(10)
	ll.print()
	ll.remove(-1)
	ll.print()
}
